components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90

```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
+                        conu,           dconudt_activa          ) ! inout
+
+            ! wet removal
+            call compute_wetdep_tend(                              &
+                doconvproc_extd,        icol,   kk,     dt,        & ! in
+                dt_u,   dp_i,           cldfrac_i,      mu_p_eudp, & ! in
+                aqfrac, icwmr,          rprd,           clw_cut,   & ! in
+                conu,                   dconudt_wetdep             ) ! inout
+
+            ! compute updraft fractional area; for compute_fluxes_tendencies use
+            ! *** these must obey  dt_u(k)*mu_p_eudp(k) = dpdry_i(k)*fa_u(k)
+            fa_u(kk) = dt_u(kk)*(mu_p_eudp(kk)/dpdry_i(kk))
+
+       endif    ! "(mu_p_eudp(k) > mbsth)"
+   enddo  ! "kk = kbot, ktop, -1"
+
+   end subroutine compute_updraft_mixing_ratio
+
+!====================================================================================
+   subroutine compute_wup(                              &
+                        iconvtype,      kk,     mu_i,   & ! in
+                        cldfrac_i,   rhoair_i, zmagl,   & ! in
+                        wup                             ) ! inout
+!-----------------------------------------------------------------------
+! estimate updraft velocity (wup)
+! do it differently for deep and shallow convection
+!-----------------------------------------------------------------------
+   use ppgrid, only: pver, pverp
+
+   integer,  intent(in) :: iconvtype            ! 1=deep, 2=uw shallow
+   integer,  intent(in) :: kk                   ! vertical level index
+   real(r8), intent(in) :: mu_i(pverp)          ! mu at current i (note pverp dimension) [mb/s]
+   real(r8), intent(in) :: cldfrac_i(pver)      ! cldfrac at current icol (with adjustments) [fraction]
+   real(r8), intent(in) :: rhoair_i(pver)       ! air density at current i [kg/m3]
+   real(r8), intent(in) :: zmagl(pver)          ! height above surface [m]
+   real(r8), intent(inout)      :: wup(pver)    ! mean updraft vertical velocity at current level updraft [m/s]
+
+   integer      :: kp1                  ! kk + 1
+   real(r8)     :: zkm                  ! height above surface [km]
+   real(r8), parameter  :: w_peak = 4.0_r8      ! pre-defined peak updraft [m/s]
+
+! shallow - wup = (mup in kg/m2/s) / [rhoair * (updraft area)]
+    if (iconvtype /= 1) then
+       wup(kk) = (mu_i(kp1) + mu_i(kk))*0.5_r8*hund_ovr_g &
+                      / (rhoair_i(kk) * (cldfrac_i(kk)*0.5_r8))
+       wup(kk) = max( 0.1_r8, wup(kk) )
+
+! deep - the above method overestimates updraft area and underestimate wup
+!    the following is based lemone and zipser (j atmos sci, 1980, p. 2455)
+!    peak updraft (= 4 m/s) is sort of a "grand median" from their GATE data
+!       and Thunderstorm Project data which they also show
+!    the vertical profile shape is a crude fit to their median updraft profile
+   else
+       zkm = zmagl(kk)*1.0e-3
```

Comment:
We should eventually factor out unit conversions. For example, here we can declare`m_to_km = 1.0e-3` and use it for unit conversion.